src/wasm/wasm-binary.cpp

```diff
@@ -1408,7 +1419,19 @@ HeapType WasmBinaryBuilder::getHeapType() {
 }
 
 Field WasmBinaryBuilder::getField() {
-  auto type = getConcreteType();
+  // The value may be a general wasm type, or one of the types only possible in
+  // a field.
+  auto initial = getS32LEB();
+  if (initial == BinaryConsts::EncodedType::i8) {
+    auto mutable_ = getU32LEB();
+    return Field(Field::i8, mutable_);
+  }
+  if (initial == BinaryConsts::EncodedType::i16) {
+    auto mutable_ = getU32LEB();
+    return Field(Field::i16, mutable_);
+  }
```

Comment:
Any reason for special-casing these? Can't we include `i8` and `i16` in `getType` too?

Reply:
`i8` and `i16` are only valid in a `Field`. We have to actually call a different constructor of `Field` when using one of them, as opposed to a normal wasm Type. And similarly `Type` can't contain them.

In theory though this could have all been different, I think. Type could have contained `i8, i16` and the validator would check those types are only used where allowed. I'm not sure offhand if that would be better though overall.

Reply:
Ah right, I missed the type of the argument to `Field` constructor is different here.